apps/elasticsearch.go

```diff
@@ -0,0 +1,196 @@
+package apps
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+)
+
+type LoggingProcessorElasticsearchJson struct {
+	confgenerator.ConfigComponent `yaml:",inline"`
+}
+
+func (LoggingProcessorElasticsearchJson) Type() string {
+	return "elasticsearch_json"
+}
+
+func (p LoggingProcessorElasticsearchJson) Components(tag, uid string) []fluentbit.Component {
+	c := []fluentbit.Component{}
+
+	// When Elasticsearch emits stack traces, the json log may be spread across multiple lines,
+	// this parser handles that case.
+	multilineParser := &confgenerator.LoggingProcessorParseMultiline{
+		Rules: []confgenerator.MultilineRule{
+			{
+				StateName: "start_state",
+				NextState: "cont",
+				Regex:     `^{.*`,
+			},
+			{
+				StateName: "cont",
+				NextState: "cont",
+				Regex:     `^[^{].*[,}]$`,
+			},
+		},
+	}
+
+	jsonParser := &confgenerator.LoggingProcessorParseJson{
+		ParserShared: confgenerator.ParserShared{
+			TimeKey:    "timestamp",
+			TimeFormat: "%Y-%m-%dT%H:%M:%S,%L%z",
+		},
+	}
+
+	c = append(c, multilineParser.Components(tag, uid)...)
+	c = append(c, jsonParser.Components(tag, uid)...)
+	c = append(c, p.severityParser(tag, uid)...)
+	c = append(c, p.nestingProcessors(tag, uid)...)
+
+	return c
+}
+
+func (p LoggingProcessorElasticsearchJson) severityParser(tag, uid string) []fluentbit.Component {
+	severityKey := "logging.googleapis.com/severity"
+	return fluentbit.TranslationRenameComponents(tag, "level", severityKey, []struct {
+		SrcVal  string
+		DestVal string
+	}{
+		{
+			SrcVal:  "TRACE",
+			DestVal: "DEBUG",
+		},
+		{
+			SrcVal:  "DEBUG",
+			DestVal: "DEBUG",
+		},
+		{
+			SrcVal:  "INFO",
+			DestVal: "INFO",
+		},
+		{
+			SrcVal:  "WARN",
+			DestVal: "WARNING",
+		},
+		{
+			SrcVal:  "DEPRECATION",
+			DestVal: "WARNING",
+		},
+		{
+			SrcVal:  "ERROR",
+			DestVal: "ERROR",
+		},
+		{
+			SrcVal:  "CRITICAL",
+			DestVal: "ERROR",
+		},
+		{
+			SrcVal:  "FATAL",
+			DestVal: "FATAL",
+		},
+	})
+}
+
+func (p LoggingProcessorElasticsearchJson) nestingProcessors(tag, uid string) []fluentbit.Component {
```

Comment:
This was something that was still being discussed last I heard;

Was any decision reached about whether we want to nest dotted keys or not?

Reply:
Let's make them nested for easier filter syntax in Cloud Logging UI.

Reply:
Sounds good, that's what's being done right now.